scripts/py_matter_yamltests/matter_yamltests/pics_checker.py

```diff
@@ -0,0 +1,139 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an 'AS IS' BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import unicodedata
+
+COMMENT_CHARACTER = '#'
+VALUE_SEPARATOR = '='
+# The only correct value to enable a feature is '1'. All others turns it off.
+VALUE_ENABLED = '1'
+
+
+class PICSChecker():
+    __pics: None
+    __index: 0
+
+    def __init__(self, pics_file: str):
+        if pics_file is not None:
+            self.__pics = self.__parse(pics_file)
+
+    def check(self, pics):
+        if pics is None:
+            return True
+
+        self.__index = 0
+        tokens = self.__tokenize(pics)
+        return self.__evaluate_expression(tokens, self.__pics)
+
+    def __parse(self, pics_file: str):
+        pics = {}
+        with open(pics_file) as f:
+            line = f.readline()
+            while line:
+                line = self.__clear_input(line)
+                if len(line) and line[0] != COMMENT_CHARACTER:
+                    key, value = line.split(VALUE_SEPARATOR)
+                    pics[key] = value == VALUE_ENABLED
+                line = f.readline()
+        return pics
+
+    def __evaluate_expression(self, tokens: list[str], pics: dict):
+        leftExpr = self.__evaluate_sub_expression(tokens, pics)
+        if self.__index >= len(tokens):
+            return leftExpr
+
+        if tokens[self.__index] == ')':
+            return leftExpr
+
+        token = tokens[self.__index]
+
+        if token == '&&':
+            self.__index += 1
+            rightExpr = self.__evaluate_sub_expression(tokens, pics)
+            return leftExpr and rightExpr
+
+        if token == '||':
+            self.__index += 1
+            rightExpr = self.__evaluate_sub_expression(tokens, pics)
+            return leftExpr or rightExpr
+
+        raise KeyError(f'Unknown token: {token}')
+
+    def __evaluate_sub_expression(self, tokens: list[str], pics: dict):
+        token = tokens[self.__index]
+        if token == '(':
+            self.__index += 1
+            expr = self.__evaluate_expression(tokens, pics)
+            if tokens[self.__index] != ')':
+                raise KeyError('Missing ")"')
+
+            self.__index += 1
+            return expr
+
+        if token == '!':
+            self.__index += 1
+            expr = self.__evaluate_expression(tokens, pics)
+            return not expr
+
+        token = self.__normalize(token)
+        self.__index += 1
+
+        if pics.get(token) == None:
+            # By default, let's consider that if a PICS item is not defined, it is |false|.
+            # It allows to create a file that only contains enabled features.
+            return False
+
+        return pics.get(token)
+
+    def __tokenize(self, expression: str):
```

Comment:
I don't know for sure. I "guess" this is used by the test harness team to know if they should run a specific yaml file based on some pics code. I have not runned the PICSChecker on it because the values that are used there are not in our default pics configuration for CI and so we won't run those. We would need to update that.

Reply:
@krypton36 ?

Reply:
This is to handle scenarios where test require certain pics and not the others. For example there is a YAML file that has the following:
PICS: "(WIFI || ETH) && !THREAD"

If the device being tested has WIFI or ETH and not THREAD, then this test will run. This tokenizer appears to tokenize the operators for these particular cases.